### PR TITLE
Fix Issue 13877 - join assumes forward range can be cheaply iterated twice

### PR DESCRIPTION
**Pull request by Poita**
e.g. with `r.map!(someExpensiveFunction).join`, `join` would previously iterate twice: once to compute length, and again to build up the result. The extra iteration to compute length may be disproportionately expensive, so we now only precompute length for built-in arrays, since we know those are cheap to iterate.

The heuristic could be improved over time, although I can't think of anything better right now.

https://issues.dlang.org/show_bug.cgi?id=13877

### CODE REVIEW

File: std/array.d [hunk 1]
```diff
@@ -1750,6 +1759,27 @@ unittest
     assert([[tuple("x")]].join == [tuple("x")]);
 }
 
+// Issue 13877
+unittest
+{
+    // Test that the range is iterated only once.
+    import std.algorithm : map;
+    int c = 0;
+    auto j1 = [1, 2, 3].map!(_ => [c++]).join;
```
Comment (AndrejMitrovic) [inline std/array.d]: `c++` => deliberate pun? :P

**Comment by AndrejMitrovic** [comment]
LG for me.

**Comment by Safety0ff** [comment]
LGTM.

**Comment by DmitryOlshansky** [comment]
LGTM though it's unclear how are we going to expand that heuristic?

It's awful to keep piling up specialization of built-ins only.

P.S. Rebase please to nudge auto-tester.

**Comment by Safety0ff** [comment]
> unclear how are we going to expand that heuristic?

I think the only option is to use tags or UDAs to provide the information needed to decide whether or not to enable the optimization.

**Comment by Poita** [comment]
> LGTM though it's unclear how are we going to expand that heuristic?

How about:

```d
private template hasCheapIteration(R) {
  static if (is(typeof(R.hasCheapIteration)))
    enum hasCheapIteration = R.hasCheapIteration;
  else
    enum hasCheapIteration = isArray!R;
}
```
Range authors then just need to add a `enum hasCheapIteration = true;` member to their types.

**Comment by DmitryOlshansky** [comment]
@Poita Sounds sensible in isolation. However it's a question of coming up with solid document like DIP that lays out all  of extra capabilities/hints of a range. I mean - how are we going to make all users know what extras a range should or shouldn't define and what extra tests a library writer should use?

In fact I see a problem even with this pull - a `map` may or may not be cheap to iterate depending on the exact contents of an argument. There is no way to determine it statically and generically.

**Comment by Poita** [comment]
@DmitryOlshansky: that's true, although we can easily add them as necessary.

You are right about `map` not necessarily being expensive. This is a general problem with conservative heuristics. I think it is better to assume the map is expensive and pay the cost of using Appender than to assume the opposite and pay an arbitrarily high cost.

**Comment by DmitryOlshansky** [comment]
As far as heuristic goes - I have an idea:  a range having L-value elements most likely is  cheap, R-value may be result of computation but L-value implies having a slot in memory.

**Comment by Poita** [comment]
I think that will work most of the time, but it's very possible to have an expensive computation of an L-value. False negatives are ok, but I don't think it would be wise to admiss false positives.

**Comment by DmitryOlshansky** [comment]
> but it's very possible to have an expensive computation of an L-value

I'd need an example of that. The things I can think of:
a) modifying L-value in place, which is something map better not do anyway
b) allocating new stuff and returning a `ref` to it, this is might be more sensible. But, for instance, a new-ed class reference is an r-value, same with sturct. To get `ref` would need more work to dereference it so I'd be hard-pressed to see anything practical.

In contrast L-value heuristic would admit ranges over containers not only arrays. It's still quite different to "cheap to iterate" though.

>  False negatives are ok

Not quite - accumulating rare special cases that only ever work with certain kind of built-in is  extra baggage that make everything harder to maintain for marginal gain.

**Comment by quickfur** [comment]
Why can't we check `hasLength` instead of hard-coding built-in arrays as a special case?

**Comment by DmitryOlshansky** [comment]
@quickfur It already had the check for hasLength.

**Comment by Poita** [comment]
@DmitryOlshansky: I'm thinking of something like `map!(x => arr[f(x)])`, where `arr` is an array, and `f` is an arbitrarily expensive function.

**Comment by DmitryOlshansky** [comment]
@Poita While a good example it's quite opposite of typical code - I mean complex table lookups may calculate some "hash" twice but is it more costly then reallocating array (potentially multiple times) or overallocating?

Either way I understand that we can't have any realistically good heuristic that isn't immediately destroyed by `map` and the like.

One thing I'm not so sure about is - this stuff will pessimize some code and optimize other code and there is nothing a user could possibly do about it. Who is going to judge what is "faster" for everybody?

**Comment by andralex** [comment]
Auto-merge toggled on

**Comment by andralex** [comment]
Nice work and idea for isCheapIteration, thanks!

**Pull request merged**